File: src/zoneA/BallQueue.ts
export class BallQueue {
  private minTier = 1;
  private maxTier = 4;
  private seeded: number[] = [];

  private currentTier: number;
  private nextTier: number;

  constructor() {
    this.currentTier = this.pullNext();
    this.nextTier = this.pullNext();
  }

  /** Tier of the ball ready to drop now. */
  peek(): number { return this.currentTier; }

  /** Tier of the ball after the current one (drives the preview). */
  peekNext(): number { return this.nextTier; }

  /** Consume the current ball, advance the queue, and return the dropped tier. */
  pop(): number {
    const dropped = this.currentTier;
    this.currentTier = this.nextTier;
    this.nextTier = this.pullNext();
    return dropped;
  }

  /** Update the random tier range. Takes effect for any future random draws. */
  setWindow(minTier: number, maxTier: number): void {
    this.minTier = minTier;
    this.maxTier = maxTier;
  }

  /**
   * Re-draw the current + next tiers from the active window. Called when a milestone shifts the
   * window up so the in-hand and preview balls aren't stranded on now-blacklisted low tiers
   * (setWindow alone only governs *future* draws). Clears any unconsumed seed first, so the
   * draw always comes from the live random window — never a stale low-tier seed.
   */
  reroll(): void {
    this.seeded = [];
    this.currentTier = this.pullNext();
    this.nextTier = this.pullNext();
  }

  /**
   * Pre-load specific tiers into the front of the queue.
   * Called on level transitions that have a defined `bufferBalls` list.
   * Once the seeded tiers are consumed, future draws use the random window.
   */
  seed(tiers: number[]): void {
    this.seeded = [...tiers];
    this.currentTier = this.pullNext();
    this.nextTier = this.pullNext();
  }

  private pullNext(): number {
    if (this.seeded.length > 0) return this.seeded.shift()!;
    return this.randomTier();
  }

  private randomTier(): number {
    const span = this.maxTier - this.minTier + 1;
    return this.minTier + Math.floor(Math.random() * span);
  }
}

File: Assets/Game/Core/BallQueue.cs
using System;
using System.Collections.Generic;

namespace RichCoast.Core
{
    /// <summary>
    /// The current + next ball draw (port of the Phaser <c>BallQueue.ts</c>). Random draws come
    /// from the active [min,max] tier window; stages can seed a curated hand that is consumed
    /// first. The RNG is injectable so tests are deterministic.
    /// </summary>
    public sealed class BallQueue
    {
        readonly Func<int, int, int> randomInclusive;
        readonly Queue<int> seeded = new Queue<int>();
        int minTier = 1;
        int maxTier = 4;

        public int CurrentTier { get; private set; }
        public int NextTier { get; private set; }

        /// <param name="randomInclusive">Returns a tier in [min, max]. Defaults to <see cref="System.Random"/>.</param>
        public BallQueue(Func<int, int, int> randomInclusive = null)
        {
            if (randomInclusive == null)
            {
                var rng = new Random();
                randomInclusive = (lo, hi) => rng.Next(lo, hi + 1);
            }
            this.randomInclusive = randomInclusive;
            CurrentTier = PullNext();
            NextTier = PullNext();
        }

        /// <summary>Consume the current ball, advance the queue, and return the dropped tier.</summary>
        public int Pop()
        {
            int dropped = CurrentTier;
            CurrentTier = NextTier;
            NextTier = PullNext();
            return dropped;
        }

        /// <summary>Update the random tier range. Takes effect for future random draws only.</summary>
        public void SetWindow(int min, int max)
        {
            minTier = min;
            maxTier = max;
        }

        /// <summary>
        /// Re-draw current + next from the live window (a milestone shifted it up, so the in-hand
        /// and preview balls aren't stranded on blacklisted tiers). Clears any unconsumed seed.
        /// </summary>
        public void Reroll()
        {
            seeded.Clear();
            CurrentTier = PullNext();
            NextTier = PullNext();
        }

        /// <summary>Pre-load specific tiers into the front of the queue; random draws resume after.</summary>
        public void Seed(IEnumerable<int> tiers)
        {
            seeded.Clear();
            foreach (var t in tiers) seeded.Enqueue(t);
            CurrentTier = PullNext();
            NextTier = PullNext();
        }

        int PullNext() => seeded.Count > 0 ? seeded.Dequeue() : randomInclusive(minTier, maxTier);
    }
}
